File: 2025/day3.py
from utils import read_input_file


def solve_part1(banks: list[str]) -> int:
    """
    Each bank consists of a string a digits. For each bank, find the largest two
    digit number that can be made by choosing two digits from the bank, preserving
    their order.  Return the sum of these numbers.
    """

    answer = 0

    for bank in banks:
        first_digit = 0
        second_digit = 0

        for i, digit in enumerate(map(int, bank)):
            if i < (len(bank) - 1) and digit > first_digit:
                first_digit = digit
                second_digit = 0
            else:
                second_digit = max(second_digit, digit)

        answer += first_digit * 10 + second_digit

    return answer


def main():
    banks = list(read_input_file("day3-input.txt").splitlines())

    part1_solution = solve_part1(banks)
    print(f"Part 1 solution = {part1_solution}")


if __name__ == "__main__":
    main()

File: 2025-exs/day3.exs
defmodule Day3 do
  def parse_input(txt) do
    String.split(txt)
  end

  @doc """
  Bank is a list of digits, represented as strings.  Returns the largest
  number that can be made by choosing num_digits digits from bank, preserving
  their order.
  """
  def solve_bank(bank, num_digits) do
    bank_len = String.length(bank)

    String.graphemes(bank)
    |> Enum.with_index()
    |> Enum.reduce(List.duplicate("0", num_digits), fn {bank_digit, i}, digits ->
      num_left = bank_len - i

      Enum.reduce_while(Enum.with_index(digits), [], fn {digit, j}, new_digits ->
        if j < num_digits - num_left or bank_digit <= digit do
          {:cont, [digit | new_digits]}
        else
          {:halt, List.duplicate("0", num_digits - j - 1) ++ [bank_digit | new_digits]}
        end
      end)
      |> Enum.reverse()
    end)
    |> Enum.join()
    |> String.to_integer()
  end

  @doc """
  Each bank consists of a string a digits. For each bank, find the largest
  number that can be made by choosing num_digits digits from the bank, preserving
  their order.  Return the sum of these numbers.
  """
  def solve(banks, num_digits) do
    Enum.sum_by(banks, &solve_bank(&1, num_digits))
  end

  def solve_part1(banks) do
    solve(banks, 2)
  end

  def solve_part2(banks) do
    solve(banks, 12)
  end
end

Aoc2025.Utils.run_day(
  "day3-input.txt",
  &Day3.parse_input/1,
  &Day3.solve_part1/1,
  &Day3.solve_part2/1
)
